add user doc for the shared file system (#79)
ChameleonCloud/docs: source/technical/shares.rst

.. _shares:

====================
Shares
====================

Chameleon provides a shared file system service through the `OpenStack Manila <https://docs.openstack.org/manila/latest/>`_ interface.
With the service, you can create a shared file system, mount to the bare metal instances, and manage some of its properties, such as visibility.

  .. hint::
  
    Chameleon shared file system service is currently backed by a CephFS. Same as our :ref:`object store <object-store>` service, the data is
    replicated and the replication should provide good availability in case of hardware failures.

    *Difference between shared file system and object store*
	You can choose either shared file system or object store to store, manage, and share your data with your collaborators. The object store
	is suitable for storing large objects at scale, but isn’t suitable for transactional data, as objects are immutable and updated in their

> 	entirety. On the other hand, Chameleon shared file system maps the CephFS using file-based NFS protocol to the bare metal instances, and
> 	the NFS protocol manages locking and data integrity processes required to provide multiple concurrent access to data. 

Reviewer: ```suggestion
	entirety. Chameleon shared file system instead provides a NFS mount to the bare metal instances, with
	the NFS protocol managing locking and data integrity processes required to provide multiple concurrent access to data.
```

Reviewer: Kind of a nit, "on the other hand" implies the sentence is going to have the same structure (describing access patterns and how writes work), but it describe slightly different aspects of the solution.

The shared file system service is available at `CHI@UC <https://chi.uc.chameleoncloud.org/>`_ and `CHI@TACC <https://chi.tacc.chameleoncloud.org/>`_.
Each region has its own service and the shares created at one region are not available to the other. As all other Chameleon services, you can create
and manage your shares using both GUI and CLI.

> .. _storage_vlans:
> 
> Storage VLANs
> ================================
> 
> To provide isolation among shares created by different projects, accessing a share requires the storage VLANs, which are special VLANs you can
> reserve to use. When reserving the storage VLANs, add `usage_type=storage` to the resource properties. To learn more about reserving VLANs, read
> the :ref:`reservations documentation <reservations>`. All bare metal instances that are created on the storage VLAN have access to all the project
> shares.

Reviewer: ```suggestion
.. _storage_network:

Storage Networks
================================

To provide isolation among shares created by different projects, accessing a share requires a storage network, which are special networks you can
reserve to use. When reserving a storage network, add `usage_type=storage` to the resource properties. To learn more about reserving networks, read
the :ref:`reservations documentation <reservations>`. All bare metal instances that are created on the storage network have access to all the project
shares.
```

I think we refer to reservable networks as networks and not VLANs, I think just saying "network" gets the point across and is less technical for, e.g., storage researchers who many not know what a VLAN is or how it relates.

Shares
================================

Visibility
--------------------------------

Shares are owned by the project. By default, all shares have `private` visibility and can only be listed and accessed within your project.

> All bare metal instances owned by the project have the read and write permissions to the project’s shares. You can also make your shares `public`.

Reviewer: ```suggestion
All bare metal instances owned by the project have read and write permissions to the project’s shares. You can also make your shares `public`.
```

Reviewer: This is a tricky thing in English... there are many cases where an article ("the", "a"/"an") is not used. I tried to find a summary of the cases but as with many things in English, it's very nonstandard :(

https://www.grammarly.com/blog/articles/

<img width="717" alt="Screen Shot 2022-05-20 at 9 49 27 AM" src="https://user-images.githubusercontent.com/796413/169554082-c287abf5-d2b0-47ca-85f1-5353fc9cbc1d.png">

The advice about not using it for "abstract ideas" I think is a general rule, at least I can't think of exceptions. "read and write permissions" are an abstract idea: you can't hold read and write permissions, they are not an object or thing.

All Chameleon users and projects can list public shares, and with a storage VLAN, all projects have read-only access to a public share.

Accessibility
--------------------------------

A share is a pre-allocated storage space at a CephFS. You can :ref:`mount your shares to your bare metal instances via NFS protocol <mount-share>`.
The accessibility of the shares are controlled internally by the reservation service. You are not allowed to edit the access rules of a share.

Quotas
--------------------------------

> We do not charge SUs for the storage spaces of your shares. However, we do limit the total size and the number of shares you can create within

Reviewer: "the storage spaces" - "the" is actually appropriate here, at least to me. So... it's a hard thing to generalize I guess.

your project. The maximum number of shares is 10 and the maximum size allowed for all shares in a project is 2000 GiB. If you need to increase
the default quota, please submit a ticket via the `Help Desk <https://chameleoncloud.org/user/help/>`_.

Managing Shares using GUI
================================

To manage your share, use the `Shares` page at `CHI@UC <https://chi.uc.chameleoncloud.org/>`_ or `CHI@TACC <https://chi.tacc.chameleoncloud.org/>`_
by navigating to `Project > Share > Shares`.

  .. figure:: shares/sharespage.png
      :alt: The Shares page

      The Shares page

Create Share
--------------------------------

Click the `Create Share` button. In the `Create Share` dialog, provide a name and the size of your share, and then click the `Create` button to
create a share.

  .. figure:: shares/createshare.png
      :alt: The Create Share dialog

      The Create Share dialog

  .. note::
  

>     The storage VLAN is not required for creating shares. It’s only required to access the shares.

Reviewer: ```suggestion
    A storage network is not required for creating shares. It’s only required to access the shares.
```

.. _view-share-gui:

View Share
--------------------------------

You can look at the details of a share by clicking the share name in the `Shares` page. Note that the paths of the `export locations` are important
as you will use this path to mount your share to your bare metal instances. You can also see the other properties, such as visibility and size.
The access rules are listed in the `share details` page, though you can not edit the rules, as they are controlled by the reservation service.

  .. figure:: shares/sharedetails.png
      :alt: The Share details

      The Share details
      
Edit Share
--------------------------------
You can manage the properties and extend the size of a share by clicking the `Action` dropdown in the `Shares` page. 

  .. figure:: shares/manageshare.png
      :alt: The Action dropdown

      The Action dropdown

Delete Share
--------------------------------
You can use the `Action` dropdown to delete a single share, or select multiple shares and click the `Delete Shares` button. 

  .. important::
  

>     Be careful when deleting shares, as the action is irreversible. However, the termination of your VLAN reservation **DOES NOT** delete your share.

Reviewer: ```suggestion
    Be careful when deleting shares, as the action is irreversible. However, the termination of your storage network reservation **DOES NOT** delete your share.
```

    Your shares persist until you manually delete them.

Managing Shares using CLI
================================

As all other Chameleon services, you can manage your shares via CLI as well. 

  .. tip::
  
    Reading :ref:`Command Line Interface (CLI) <cli>` is highly recommended before continuing on the following sections.

In addition to installing the CLI, you must also install `python-manilaclient` package:

  .. code-block:: bash

    pip install python-manilaclient

Then, you must set environment variables for your account and project using :ref:`The OpenStack RC Script <cli-rc-script>`.

List Shares
--------------------------------

To list all shares of your project, run the following command:

  .. code-block:: bash

>     manila list

Reviewer: In general I think we should prefer to use the `openstack` flavor of commands, is there one for manila?

For example, for creating a 1 GiB share with name of ``my-first-share``, run:

  .. code-block:: bash
  
    manila create --name my-first-share NFS 1

  .. note::
 
    Only the NFS protocol is supported. 

You can add the ``--public`` flag to make your share public.

Edit Share
--------------------------------

To change the visibility of a share, run:

  .. code-block:: bash

    manila update --is-public <true/false> <name/id of the share>

To update the name or the description of a share, run:

  .. code-block:: bash

    manila update --name <new name> --description <description> <name/id of the share>

To extend/shrink the size of a share, run:

  .. code-block:: bash

    manila extend <name/id of the share> <new size>
    manila shrink <name/id of the share> <new size>

.. _view-share-cli:

View Share
--------------------------------

To view the details of a share, run:

  .. code-block:: bash

    manila show <name/id of the share>

Delete Share
--------------------------------

To delete a share, run the following command:

  .. code-block:: bash

    manila delete <name/id of the share>

.. _mount-share:

Mounting Shares to Instances
================================

> In order to allow your instances to talk to the CephFS backend, you need to create your instances using the :ref:`pre-reserved storage VLAN <storage_vlans>`.

Reviewer: ```suggestion
In order to allow your instances to access the share, you need to create your instances using the :ref:`pre-reserved storage network <storage_networks>`.
```

I don't think we need to say how it is implemented

To learn more about how to create a bare metal instance on a network, read :ref:`the bare metal instances section <baremetal-gui-launch>`. 

  .. important::

>     The shares are independent of the storage VLANs. You can create shares any time regardless of the status of the storage VLANs. The storage VLANs are

Reviewer: ```suggestion
    The shares are independent of the storage networks. You can create shares any time regardless of the status of the storage networks. The storage networks are
```
